### ISSUE TITLE
Add a maximal speed in which the doors and baggage compartment door can be opened

### ISSUE THREAD

**Issue by gilbertohasnofb** [c172p-team/c172p#544]
Add a maximal speed in which the doors and baggage compartment door can be opened. Currently they can be opened at any speed and in-flight.

**Comment by onox** [comment]
I would like to implement this as just limiting how much the doors can be opened?

**Comment by onox** [comment]
I would like to add a wind force that pushes against the doors.

**Comment by gilbertohasnofb** [comment]
:+1:

**Comment by gilbertohasnofb** [comment]
As for the baggage door, the PR https://github.com/Juanvvc/c172p-detailed/pull/546 will not allow it to be opened mid-flight any more. But if one left it opened and took off, the door will remain opened, so if you want to improve this via wind forces it could be a nice addition. Else it is a minimal detail anyway :smile:

**Comment by dg-505** [comment]
You may want to take a look at the DHC6 Twin Otter. There the doors can be opened until a specific speed. Maybe this helps you implemeting that. Just a suggestion...

**Comment by onox** [comment]
Baggage compartment door can only be opened on ground, but could be left opened while the aircraft is taxiing or taking off.

**Comment by gilbertohasnofb** [comment]
@onox great idea, and this is another good bug for beginners.

**Comment by onox** [comment]
I think this could be implemented with a badly PID controller. The pilot provides a "force" as well as the wind, this is then used to compute the maximum open position of a door. This is then used as a set point that is fed to the PID controller. The controller provides a smooth animation to the set point (the maximum position). If badly tuned then it should oscillate a bit, simulating the pilot fighting against the wind.

**Comment by wkitty42** [comment]
On 03/10/2018 07:39 AM, onox wrote:
> Baggage compartment door can only be opened on ground, but could be left opened 
> while the aircraft is taxiing or taking off.


instead of setting a speed, how about tracking the wind speed and slamming the 
door shut with a nice loud bang which will get the pilot's attention :evilgrin:

**Comment by gilbertohasnofb** [comment]
@onox limiting the max position allowed at certain airspeeds is easy as one just need to use a multiplier proportional to the inverse of the airspeed within a certain range. But I don't understand what you wrote about using the PID controller to animate the oscillations of an open door in the wind. Isn't there any other way of implementing a smooth random fluctuation to a property?

**Comment by gilbertohasnofb** [comment]
@wlbragg it seems we are using a function called `toggle` to animate the doors and windows from a position of `0.0` into a position of `1.0` using the following Nasal code: `c172p.rightWindow.toggle();`

So let's suppose I create a filter which will take the value of `/sim/model/door-positions/rightWindow/position-norm` (0.0 to 1.0) and, using a multiplier proportional to the aispeed, will spit out a new property `/sim/model/door-positions/rightWindow/position-norm-effective` which will range from (0.0 to a value less than or equal to 1.0, depending on the airspeed). So this is all good, the function `c172p.rightWindow.toggle();` can still be called from the pick animation since we still want `position-norm` to toggle between `0.0` and `1.0` but how do I use `position-norm-effective` instead of `position-norm` in the animation of the doors? Do you know where those animations are defined? I can't find them...

**Comment by wkitty42** [comment]
On 07/22/2018 12:27 PM, wlbragg wrote:
> @gilbertohasnofb <https://github.com/gilbertohasnofb> I don't suppose your going 
> to try to add drag to the FDM for an open door? For one you probably can't get 
> it open at any kind of speed, enough to produce much drag, right? Also the 
> "window open" drag is probably negligible as well and not worth adding?


i agree with this... i wouldn't worry about drag at all for these features (open 
door, open window) at all...

**Comment by legoboyvdlp** [comment]
Actually, if you have ever driven down a highway with a window open (considering the Cessna flies at 130mph or so) I would guess the window drag would actually be fairly large - but thats just a guess. If there is any actual data available anywhere to simulate this, I think it would be good to simulate it.

**Comment by wkitty42** [comment]
On 07/22/2018 01:21 PM, Jonathan Redpath wrote:
> Actually, if you have ever driven down a highway with a window open
> (considering the Cessna flies at 130mph or so) I would guess the window drag
> would actually be fairly large - but thats just a guess. If there is any
> actual data available anywhere to simulate this, I think it would be good to
> simulate it.

there's this chart i found on the wiki while looking for something else...

   http://wiki.flightgear.org/AI_Systems#Using_Interpolation_Tables

it'll be handy in another FG related project... i'm sad that it doesn't specify 
the altitude for the given speed, though... there's other charts, too... maybe 
we need to do some of these on the wiki page so we have more complete data 
available that's easy to find?

the window in a car is a lot different... it rolls down into the door... the 
windows in the c172p just pop out at the bottom a few inches... not a very large 
area for the wind to be coming into the craft and hitting the rear cabin wall...

an open door is different, too... by the time you're doing 10kts, these 
free-swinging doors should be already closed... unlike a car's doors that have 
detentes on them to hold them open at one of two positions... they may hold 
until 15mph but not very well in my experience...

it may be worth it but is the pain:gain ratio within reason...

**Comment by gilbertohasnofb** [comment]
I had implemented drag on my branch but I can remove it if you guys don't want it. But the drag is proportional to the door position, so if the door open just partially in flight then the drag will be partial as well. Right now the drag is perceptible only when the doors are fully opened. But again, I can remove it if you guys want. Once the branch is done I will put it up for you all to test.

As for the window, I am 100% sure they can be opened in flight from my single flight hour.

@wlbragg thanks for the information, I will see if I manage from here.

**Comment by gilbertohasnofb** [comment]
@wlbragg If I understand correctly, the doors are being rotated 70 degrees using a hardcoded value:
```
            <entry>
                <ind>1</ind>
                <dep>-70.0</dep>
            </entry>
```

Could I replace this value with a changeable property which would vary from 70 degrees to a much lower value when the airspeed increases?

**Comment by gilbertohasnofb** [comment]
But that would only change the maximal parameter of the animation once the door animation is activated, but would not change the position of an open door if the airspeed starts to increase... I am thinking I simply don't know enough about animations in order to tackle this one. I can gladly create that filter I mentioned which would spit a new property that will range from `0.0` to a value `<= 1.0` depending on airspeed but I am afraid I won't be able to do much more after that.

**Comment by wlbragg** [comment]
@gilbertohasnofb I don't mind having drag included if it is realistic. I'm only wondering if it could in reality be applied. Can you get up to takeoff speed with the door open or would it shut before you even could get off the ground? Could you open the door in flight and why would you?
In the Cub I add it because they actually remove the doors in some variations of service aircrafts. That is the only implementation I used in the Cub, on or off.

**Comment by wlbragg** [comment]
@gilbertohasnofb 
>Could I replace this value with a changeable property which would vary from 70 degrees to a much lower value when the airspeed increases?

I have to put some thought to the problem but I am pressed at the moment for time. As soon as I get back to my computer I will look at this and see if I have a solution.

**Comment by wlbragg** [comment]
>Could you open the door in flight and why would you?

Skydiving might be an example, do they ever jump by opening a door, or is the door removed? If they use a c172 to sky dive from and they remove the door or open it, then I would say yes, lets simulate it.

**Comment by gilbertohasnofb** [comment]
@wlbragg the idea here (if I understood correctly) is to limit the doors and baggage door opening when airspeed increase. So for instance while at 0 kts the doors can be fully opened, at 30 kts they can be only partially opened, maybe 60%. At higher airspeeds (above 70-80kts) they can be opened 5%. So the effect is that above a certain airspeed the door won't be able to be opened any longer. Makes sense?

**Comment by gilbertohasnofb** [comment]
and again, the drag will decrease with the door angle so a barely opened door will be negligible while a fully open door at 0 kts will make the aircraft take longer to accelerate.

**Comment by wlbragg** [comment]
I understand, I am conflating two ideas,..

1) Wind effect on the door being able to be opened and how much and also it should close it if to much resistance is applied from the wind.

2) Effect on the FDM if any from 1) above.

Item 1), yes we might as well simulate it. I'll try to give you some answers to the programming issues in a bit.

Item 2),  Is there ever a period for any reasonable amount of time that the door will be opened before either manually being closed or automatically being closed (from the friction) that it is worth simulating the drag on the FDM?
Along with this question comes another, do they ever remove the doors from the 172p for any reason and then fly it (sky diving maybe?). If so then we could add the drag based on the amount of door opening or removal.

I'm really just asking these questions, how far do we go with this simulation?

I'll still get back with an answer to the problem in item 1).

**Comment by gilbertohasnofb** [comment]
> how far do we go with this simulation?

IMO not too far, to be frank. A bit of drag is trivial to add but should also be so small that we can even ignore that. I think all we want is to limit the animation max angle according to the airspeed, nothing more.

**Comment by hbeni** [comment]
To propose something in between: i just sugested in the c182s implementation to just clamp the max angle of the doors.
What about adding some nasal snipped that checks airspeed every couple of seconds and trigger the door toggle in case it is open and above a certain low airspeed?

**Comment by gilbertohasnofb** [comment]
@hbeni sounds good to me. @wlbragg what do you think?

**Comment by wkitty42** [comment]
On 07/22/2018 02:21 PM, Gilberto Agostinho wrote:
> I had implemented drag on my branch but I can remove it if you guys don't
> want it.

if you've already done it, that's great! i just didn't want to see it take a 
while with all the other things going on and have it either get lost, come in 
incomplete, push the project back from the planned release branching/tagging, etc...

**Comment by wlbragg** [comment]
>To propose something in between: i just sugested in the c182s implementation to just clamp the max angle of the doors.

So clamp the angle of the door based on windspeed? Use an interpolation table?

>What about adding some nasal snipped that checks airspeed every couple of seconds and trigger the door toggle in case it is open and above a certain low airspeed?

In all honesty I think i would avoid another nasal loop checking a property every few seconds. Even though we do it, we have tried to re-factor as much of that kind of code as we could. Would a listener for airspeed with a if speed > X in it to shut the door work?
@gilbertohasnofb I would still like to investigate your approach. So I don't have to search for it, do you remember what file
~~~
<entry>
                <ind>1</ind>
                <dep>-70.0</dep>
            </entry>
~~~
was in?

**Comment by gilbertohasnofb** [comment]
Lines 3996 and 4097 of `/Models/c172p.xml`

**Comment by hbeni** [comment]
wouldn't a nasal timer-based loop executing every few seconds (like 10 or so) really be more costly as checking an xml based property every tick?

**Comment by gilbertohasnofb** [comment]
@hbeni would a 10 second loop really suffice? During take off, for instance, the speed changes very quickly so such loop must be _much_ shorter. Also, in general, I think that Nasal is always heavier than a simple xml listener.

**Comment by wlbragg** [comment]
>wouldn't a nasal timer-based loop executing every few seconds (like 10 or so) really be more costly as checking an xml based property every tick?

I don't know, maybe.  By design we have steadily removed most all the nasal loop checks. I don't know enough about the internals of a listener to know its effectiveness over a nasal loop. One thing in our case would be we add it the main c172p.nas file and it would be a couple lines of code in the listener section with all the rest VS a branched loop somewhere. I personally don't think a loop is the worst idea, I am just trying to be consistent in our approach. That is why my first choice would be that of @gilbertohasnofb original attempt at implementation because that is usually our fist choice in this project.

**Comment by legoboyvdlp** [comment]
I don't know if this is what you are suggesting, as I haven't been following the discussion, but a listener to a property changing as often as speed is a _really bad_ idea from what I know about Nasal 😉

**Comment by wlbragg** [comment]
>a listener to a property changing as often as speed is a really bad idea 

Yeah, I thought about that as well. Ideally it would be a listener for a certain speed, but that is another whole can of worms.

@gilbertohasnofb so what are you stuck on?
~~~
<property alias="../../params/doors/right/property"/>
~~~
is the alias of
~~~
/sim/model/door-positions/leftDoor/position-norm
~~~
70 is the max we are allowing the door to open with the use of the interpolation table. and the /sim/model/door-positions/leftDoor/position-norm parameter.

So a filter or whatever you had in mind that reduces /sim/model/door-positions/leftDoor/position-norm based on airspeed seems appropriate.
We then should be able to use that new property in the animation (the alias starting at lines 419 of Models/c172p.xml). We would probably have to replace /sim/model/door-positions/leftDoor/position-norm it in the multiplayer section in c172p-set line 441-442 and 443 if were doing baggage door as well with the new property.

**Comment by gilbertohasnofb** [comment]
@wlbragg A property cannot be both the input and the output of the filter. So in goes `/sim/model/door-positions/leftDoor/position-norm` and out comes `/sim/model/door-positions/leftDoor/position-norm-effective`. But what to do with that property afterwards?

**Comment by wlbragg** [comment]
I think it may be as easy as replacing
/sim/model/door-positions/leftDoor/position-norm in the multiplayer section in c172p-set line 441-442 (and 443 if were doing baggage door) 
with 
/sim/model/door-positions/leftDoor/position-norm-effective

We will have to do some MP testing to be sure all is well.

**Comment by gilbertohasnofb** [comment]
I am not sure I follow, are you suggesting that changing that multiplayer variable will affect the animation of the plane for an offline single player session? My issue is the following: toggling the doors will still toggle the `position-norm` between `0.0` and `1.0` and I can easily create a filter for an `effective` property to multiply that by a factor in relation to airspeed, but then how can I use this `effective` property to animate the door?

I think this issue is a bit beyond my capacities. I can gladly code the code for the `-effective` function so that the range is let's say `0.0`  to `1.0` at 0 kts, `0.0`  to `0.8` at 10 kts, `0.0`  to `0.4` at 30 kts and `0.0`  to `0.0` at 50 or higher kts. But then would you or anyone else be willing to do the animation bit using this property?

**Comment by wlbragg** [comment]
If I understand you correctly, yes, make the filter, I think I know what we need to do to use it.

I pushed some changes so please make sure you pull prior to coding and pushing again.

**Comment by wlbragg** [comment]
>I am not sure I follow, are you suggesting that changing that multiplayer variable will affect the animation of the plane for an offline single player session?

@gilbertohasnofb yes, that is the property that the door animations use.
/sim/model/door-positions/leftDoor/position-norm
and 
/sim/model/door-positions/rightDoor/position-norm
respectively.

In the multi-player section in the set file those properties are aliased as
~~~
<float n="0" alias="/sim/model/door-positions/rightDoor/position-norm"/>
<float n="1" alias="/sim/model/door-positions/leftDoor/position-norm"/>
~~~
Replace those with the new filtered versions and it should all work.

The animations use an alias created in Model/c172p.xml as
~~~
<params>
........
        <doors>
            <left>
                <property>sim/multiplay/generic/float[1]</property>
            </left>
            <right>
                <property>sim/multiplay/generic/float[0]</property>
            </right>
     </doors>
.........
</params>
~~~

The door animations use that alias as
~~~
<property alias="../../params/doors/leftt/property"/>
<property alias="../../params/doors/right/property"/>
~~~

**Comment by gilbertohasnofb** [comment]
@wlbragg I will try that, but at the moment I am having some issues creating my filters. Can someone help me spot my problem? Here's what I have:

```
    <filter>
        <name>Effective range for right door</name>
        <type>gain</type>
        <input>
            <product>
                <property>/sim/model/door-positions/rightDoor/position-norm</property>
                <max>
                    <min>
                        <difference>
                            <value>1.0</value>
                            <div>
                                <property>/velocities/airspeed-kt</property>
                                <value>25.0</value>
                            </div>
                        </difference>
                        <value>1.0</value>
                    </min>
                    <value>0.5</value>
                </max>
            </product>
        </input>
        <output>
            <property>/sim/model/door-positions/rightDoor/position-norm-effective</property>
        </output>
    </filter>
```

The formula is basically:

`position-norm-effective` = `position-norm` * max(0.5, (min(1.0, 1 - `airspeed` / 25)))

This should output 1.0 for 0 kts and 0.5 for anything above 50 kts. (note that we need 0.5 as our lower bound because the animation of the opening of the door starts only when `position-norm` > 0.5, lower values are used to twist the handle while the door is closed.

For some reason the resulting property is always `0.0` and I don't understand why. 

Btw I just made a test with:

```
    <filter>
        <name>Effective range for right door</name>
        <type>gain</type>
        <input>
            <value>1.0</value>
        </input>
        <output>
            <property>/sim/model/door-positions/rightDoor/position-norm-effective</property>
        </output>
    </filter>
```

and the output is 1.0, as expected. So the problem is not in the definition of the filter but on the maths I used.

**Comment by wlbragg** [comment]
@gilbertohasnofb should it be <quotient> VS<div>?

**Comment by gilbertohasnofb** [comment]
That was not the issue as changing the `<div>` to `<product>` and inverting the factor was also not giving me any result. I discovered that the issue was that I needed to wrap everything around an `<expression>` block.

**Comment by gilbertohasnofb** [comment]
> I finally figured out I needed to wrap it in the expression tag.
> That is funny we both learned that lesson at the same time. smile

@wlbragg Yeah, and I think this isn't the first time I am caught by that. I was so obsessed that there must be some maths error. Hopefully this is lesson learned this time :smile: 

> I'm close to finished with the yoke cord animation. I am trying to figure out how to stretch the model mesh.

Looking forward to this one!

**Comment by gilbertohasnofb** [comment]
@wlbragg I did find a way of incorporating `<property>/accelerations/pilot/x-accel-fps_sec</property>` but the problem is that this property is not refreshed as often as the airspeed, for instance, and it creates an ugly jittery movement. I frankly don't know of nothing else to be done...

**Comment by gilbertohasnofb** [comment]
@wlbragg ok, that property is not jittery, it was a maths problem on my side which I fixed. But after much fighting with this, I realized that this will be even harder than I had previously thought (without completely re-coding _everything_ related to doors and animations into Nasal): what you asked for in relation to the acceleration is frankly impossible since all that I am doing is modifying the factor for the animation). Thus when the acceleration reaches a certain threshold, I can try to use that value as one more factor into the door position. But as soon as the acceleration is 0, the door will then re-open to the previous value. This is because, again, I am just using factors to multiply the animation range, I am not effectively closing the door.

Also, after thinking a bit, why do you think the doors would close and not open when you hit the brakes? If the doors are not properly locked, a negative acceleration will make the doors open due to inertia + due to the wind speed being smaller and smaller. Thinking this way, our current animation is more than suitable. [Here](https://www.youtube.com/watch?v=up3W__JCp9w) is an example: open doors + extreme acceleration = the doors get closed. So the reverse must be true: improperly locked doors (near the close position) + extreme deceleration = the doors will open. 

So if you agree with this, I would like to call this done :tired_face:

**Comment by wlbragg** [comment]
>Also, after thinking a bit, why do you think the doors would close and not open when you hit the brakes?

I do think they would open if the deceleration was fast enough. But you also have wind resistance opposing against the opening at higher speeds that you don't have opposing the closing. Carefully try it in a car sometime. It is easier to get the door to close than to open. 
The effect I didn't like was the opening no matter how slow you decelerate (because it is tied the way it is using a filter). It looks unnatural in that instance. I also didn't like the direct relationship or "linear speed" of both the opening and closing, (again because it is tied to a property using a filter). I agree, no way around that using the filter except possibly a condition in the filter that won't allow opening unless a predetermined deceleration force is reached. But I think you ended up with other issues trying to do that.

So yes, I am OK with this as is. Mainly because it is a fringe case effect anyway. We can always return to it if we figure out a better way.

**Comment by gilbertohasnofb** [comment]
> I agree, no way around that using the filter except possibly a condition in the filter that won't allow opening unless a predetermined deceleration force is reached.

I still don't think you are seeing this through, there is no 'except', I frankly don't think this can be done with filters at all, condition or no condition. A filter doesn't know what you previously did to your aircraft, e.g. it doesn't know if the aircraft is stopped due to braking or just because you are simply starting the sim. If you just open the sim, you really want an open door (`position-norm == 1`) to fully open. If you just braked you want an open door (`position-norm == 1`) to be pretty much closed (`position-norm-effective == 0.51`). But there is no way of detecting any of this and smoothly transition from one case to another, there is no interpolation in filters, there is no animation in filters, etc. Frankly what you would like is a far too complex system and the pain-gain ratio is long past any reasonable value for me. If you want to tweak the system I implemented to your liking or re-code something please go ahead, I have no issues with that, but I think this is the end point for me regarding this issue. In particular since there is the new release coming up soonish, I think we should invest our time into more important issues as this one has become way too time consuming for me. I hope that's all good with you!

**Comment by wlbragg** [comment]
>no way around that using the filter except possibly a condition in the filter that won't allow opening unless a predetermined deceleration force is reached. 

>there is no 'except',

>But there is no way of detecting any of this and smoothly transition from one case to another, 

I disagree (although it might turn out in the end your correct).

> In particular since there is the new release coming up soonish, I think we should invest our time into more important issues as this one has become way too time consuming for me. I hope that's all good with you!

**Comment by hbeni** [comment]
As far as i understand, this sees like the perfect thing for nasal/xml combination, because both parts could stay pretty simple this way.

**Comment by gilbertohasnofb** [comment]
@hbeni I frankly don't think it's worth the effort. This is just a special effect, an easter egg which will hardly be noticed. Performance is really sacred and I am very reticent about using Nasal for this. But if anyone wants to give it a try and adapt or rewrite what I did I would not object that at all! I just don't think that any performance lost can be justified with an 'oh but we simulate doors banging in the wind in the most realistic way possible'.

**Comment by hbeni** [comment]
Yes, im with you.
What is the overhead of nasal? (I'm just curious as i want to learn)

**Comment by gilbertohasnofb** [comment]
I frankly don't know, I just learned the mantra that if something can be accomplished via XML then that's the better route due to performance. The performance difference might be minimal, but then again the c172p has become more and more demanding concerning performance. If this was a major feature (e.g. a system simulation in Nasal) then I would be a 100% in favour of implementing it as detailed as possible regardless of performance. But this is also just my 2 cents, other developers here might have very different opinions and I speak only for myself.

TL;DR I don't know and I might be wrong but I suspect Nasal might cause performance issues.

**Comment by wlbragg** [comment]
I think if I was to write this I would use two JSBSim fcs_functions.
One would use airspeed, the other acceleration/deceleration as conditions.
I think as conditions are met you could use interpolation tables tied to...
 function 1) = acceleration/deceleration as input with door/position-norm-effective as output
function 2) = windspeed as input with door/position-norm-effective as output

Like the filter solution, there may be issues I simply don't foresee, or I'm not accounting for, making it harder than it appears in my mind. But I think it "could" be as simple as the two functions described above.

I don't think nasal is necessary in this case. I know there has been much debate about nasal, but when used properly and when called for, it's not an issue.

**Comment by gilbertohasnofb** [comment]
> But I think it "could" be as simple as the two functions described above.

Just making it clear, please feel free to try your solution and replace mine completely if the result is more satisfactory. I wouldn't have any issues with that whatsoever!

**Comment by wlbragg** [comment]
@gilbertohasnofb I understand. Not my priority right now so I am OK with what you've implemented.
